### DISCUSSION TITLE
Forward over adjoint for Hessian

### DISCUSSION THREAD

**Discussion by Manushanker** [auto-differentiation/XAD#85] [General]
I am trying to compute the Hessian matrix for the example "samples/fwd_adj_2nd/main.cpp". The code describes how to obtain the first row of the Hessian matrix. Can you please explain how to compute the other rows of the matrix? If I change the line no 66 in the code as "derivative(value(x1_ad)) = 1.0;", then the code determines the 2nd row of the Hessian. 

Can you please explain how to obtain all the rows in the Hessian matrix?

**Comment by auto-differentiation-dev** [comment]
As the Hessian is a second order derivative, the function to be derived with forward mode is no longer a N->1 function. It is a N->N+1 function (the adjoint code for the original function, producing value and all first order derivatives). Therefore it is not possible to produce the Hessian with a single execution of the original function. 

You would need to loop over each row to produce the full Hessian, clearing the tape, resetting `derivative(value(x0_ad)) = 0.0;`, and setting `derivative(value(x1_ad)) = 1.0;`, and so on for each row.

Note that we are planning higher level functions in issue #37 to automate this task with a more convenient API. It will however execute the function multiple times as well, running the steps described above.

**Reply by Manushanker** [reply]
Thanks a lot for your comments. I will test your approach to compute the full Hessian.

**Reply by Manushanker** [reply]
I tried to follow the steps you mentioned, but not sure how to clear the tape. Can you please send me a modified main.cpp file that describes how to evaluate the full Hessian matrix? Thanks a lot!

**Reply by auto-differentiation-dev** [reply]
Hi @Manushanker,

Perhaps it’s best if you could share a code snippet of what you have done so far on the main program side, so we can take a look and see how we can help?

**Comment by auto-differentiation-dev** [comment]
For the benefit of people seeing this discussion later - there is now a helper function to compute the Hessian for a function, which does the necessary tape clearing and re-runs as needed. Please see the docs here: https://auto-differentiation.github.io/ref/hessian/
